std/range/package.d

```diff
@@ -4416,6 +4416,33 @@ body
     assert(approxEqual(rf, [0.0, 0.1, 0.2, 0.3, 0.4]));
 }
 
+///
+unittest
+{
+    import std.stdio;
+
+    /* The following groups all produce the same output of:
+     * 0 1 2 3 4
+     */
+
+    foreach (i; 0..5)
+        writef("%s ", i);
+    writeln();
+
+
+    import std.range : iota;
+    foreach (i; iota(0, 5))
+        writef("%s ", i);
+    writeln();
+
+    writefln("%(%s %|%)", iota(0, 5));
+
+    import std.algorithm : map, copy;
+    import std.format;
+    iota(0, 5).map!(i => format("%s ", i)).copy(stdout.lockingTextWriter());
```

Comment:
This fixes the trailing whitespace and does the newline in the same write-lock, and is still left to right:
`iota(0, 5).map!(to!string).joiner(" ").chain(newline).copy(stdout.lockingTextWriter);`

It doesn't fix the issue of dynamic memory allocation.